File: 90-Subsets II-子集 II/90-Subsets II-子集 II.java
class Solution {
    List<List<Integer>> res;
    LinkedList<Integer> path = new LinkedList<>();

    public List<List<Integer>> subsetsWithDup(int[] nums) {
        res = new ArrayList<>();
        Arrays.sort(nums);
        backtrace(nums, 0);
        return res;
    }

    public void backtrace(int[] nums, int index){
        if(index > nums.length) return;

        res.add(new ArrayList<Integer>(){{addAll(path);}});

        for(int i = index; i < nums.length; i++){
            if(i > index && nums[i] == nums[i - 1]) continue;

            path.add(nums[i]);
            backtrace(nums, i + 1);
            path.removeLast();
        }
    }
}

File: 90-Subsets II-子集 II/90-Subsets II-子集 II.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
 let res = []
 let path = []
 var subsetsWithDup = function(nums) {
     res = []
     nums.sort()
     backtrace(nums, 0)
     return res
 };
 
 var backtrace = function(nums, index){
     if(index > nums.length) return
     
     res.push(Array.from(path))
 
     for(let i = index; i < nums.length; i++){
         if(i > index && nums[i] === nums[i - 1]) continue
 
         path.push(nums[i])
         backtrace(nums, i + 1)
         path.pop()
     }
 }